extensions/amazon-lambda-http/runtime/src/main/java/io/quarkus/amazon/lambda/http/LambdaHttpHandler.java

```diff
@@ -46,19 +53,102 @@ public AwsProxyResponse handleRequest(AwsProxyRequest request, Context context)
+            log.error("Request Failure", e);
             return new AwsProxyResponse(500, errorHeaders, "{ \"message\": \"Internal Server Error\" }");
-        } finally {
-            connection.close();
         }
 
     }
 
-    private AwsProxyResponse nettyDispatch(VirtualClientConnection connection, AwsProxyRequest request) throws Exception {
+    private class NettyResponseHandler implements VirtualResponseHandler {
+        AwsProxyResponse responseBuilder = new AwsProxyResponse();
+        ByteArrayOutputStream baos;
+        WritableByteChannel byteChannel;
+        final AwsProxyRequest request;
+        CompletableFuture<AwsProxyResponse> future = new CompletableFuture<>();
+
+        public NettyResponseHandler(AwsProxyRequest request) {
+            this.request = request;
+        }
+
+        public CompletableFuture<AwsProxyResponse> getFuture() {
+            return future;
+        }
+
+        @Override
+        public void handleMessage(Object msg) {
+            try {
+                //log.info("Got message: " + msg.getClass().getName());
+
+                if (msg instanceof HttpResponse) {
+                    HttpResponse res = (HttpResponse) msg;
+                    responseBuilder.setStatusCode(res.status().code());
+
+                    if (request.getRequestSource() == AwsProxyRequest.RequestSource.ALB) {
+                        responseBuilder.setStatusDescription(res.status().reasonPhrase());
+                    }
+                    responseBuilder.setMultiValueHeaders(new Headers());
+                    for (String name : res.headers().names()) {
+                        for (String v : res.headers().getAll(name)) {
+                            responseBuilder.getMultiValueHeaders().add(name, v);
+                        }
+                    }
+                }
+                if (msg instanceof HttpContent) {
+                    HttpContent content = (HttpContent) msg;
+                    int readable = content.content().readableBytes();
+                    if (baos == null && readable > 0) {
+                        baos = createByteStream();
+                    }
+                    for (int i = 0; i < readable; i++) {
+                        baos.write(content.content().readByte());
+                    }
+                }
+                if (msg instanceof FileRegion) {
+                    FileRegion file = (FileRegion) msg;
+                    if (file.count() > 0 && file.transferred() < file.count()) {
+                        if (baos == null)
+                            baos = createByteStream();
+                        if (byteChannel == null)
+                            byteChannel = Channels.newChannel(baos);
+                        // AbstractNioByteChannel also seems to do a spin, albeit with a count.
+                        while (file.transferred() < file.count()) {
+                            file.transferTo(byteChannel, file.transferred());
```

Comment:
I don't think this is correct. I think you need to create a buffer that is the size of file.count(), then just calling transferTo once should be enough. If you don't have enough space in the buffer then there is no way to read all the contents.

Reply:
The channel is a ByteArrayOutputStream:
Channels.newChannel(baos);

I thought the problem would have been on the read of the file.  That's why I was worried.

Reply:
Ah yea, in that case I think it should just work without the loop. According to the docs:

`Fewer than the requested number of
     * bytes are transferred if this channel's file contains fewer than
     * <tt>count</tt> bytes starting at the given <tt>position</tt>, or if the
     * target channel is non-blocking and it has fewer than <tt>count</tt>
     * bytes free in its output buffer.
`

This should be a blocking channel, so as far as I can see it should always succeed in one call. The ones inside Netty itself are working with non-blocking channels.